### ISSUE TITLE
Game Scenes (advice)

### ISSUE THREAD

**Issue by sdfereday** [straker/kontra#115]
Hi all,
I've been looking in to using scenes in Kontra (such as different RPG areas / menu screens, etc), most of which I've got working in some manner but the method I'm currently using:

    import { init, GameLoop, load } from  "kontra";

	const { canvas } = init();

	const Scene = ({ areaId }) => {
		return GameLoop({
			update: () => console.log("Area Id:", areaId),
			render: () => {}
		});
	};

	load("assets/tileimages/test.png").then(assets => {
		const loadScene = props => {
			console.info("==> Next Scene:", props);
	
			return Scene({
				...props
			}).start();
		};

		loadScene({ areaId: "area1" });

		setTimeout(() => {
			loadScene({ areaId: "area2" });
		}, 3000);
	});

However if you run the script, it doesn't seem to clear out the 'area1' scene and keeps it in memory.  Is there a crucial point I'm missing / does anyone have experience with implementing scenes at all?

This isn't really an issue per-say, so my apologies if I've asked in the wrong bit.

Cheers

**Comment by straker** [comment]
Nice! I personally don't have any experience trying to write scenes in Kontra, but maybe @wini3d , @BoldBigflank, or @Rayne do?

Otherwise, just from the posted code, it looks like the current `GameLoop` may not be stopped when you load a new scene. Might be what's causing the previous scene from not being cleared.

**Comment by BoldBigflank** [comment]
You call `start()` but never call `stop()`, you'll need to save what the current scene is and stop it before starting the next.

**Comment by sdfereday** [comment]
Amazing, thanks so much! I’ll gjve that a go. Totally missed the stop method in my travels by overthinking the problem it seems.

**Comment by sdfereday** [comment]
I can confirm that it works a treat! Thanks for sorting me out everyone. :)
If it's of any use to folks I ended up settling for the following (for now at least) and stuck it all in a scene manager script:
```js
export default ({ scenes = [] }) => {
  let currentScene = null;

  return {
    loadScene: props => {
      const { areaId } = props;

      if (currentScene) currentScene.stop();

      const { sceneObject: Scene } = scenes.find(
        scene => scene.areaId === areaId
      );

      currentScene = Scene(props);
      return currentScene.start();
    }
  };
};
```

What a scene object might look like (different scenes will hold different properties and things):
```js
const Scene = ({ areaId }) => {
  console.info("==> Next Scene Loaded:", areaId);

  return GameLoop({
    update: () => console.log("Area Id:", areaId),
    render: () => {}
  });
};
```

How it could be used:
```js

/* Pass in some pre-existing scene data as an example */
const sceneManager = SceneManager({
    scenes: [
      {
        areaId: "area1",
        sceneObject: Scene
      },
      {
        areaId: "area2",
        sceneObject: Scene
      }
    ]
  });

/* Leave it to the scene manager to find the one it needs */
  sceneManager.loadScene({
    areaId: "area1"
  });
```

Obviously implementation is all up for debate but so far so good.

**Comment by Rayne** [comment]
Looks good to me. :+1:

I'm currently refactoring my JS13K game to support "game states" ("scenes"). My approach is similiar, but it also supports stacking scenes. Stacking can be useful for menus that are not occupying the whole screen and for reusing already loaded scenes. The top most game state is updated and rendered. If the top most game state is an overlay, then the state below will be (recursively) rendered, too.

**Workflow**

1. The intro scene is pushed to the game state stack
2. When the intro ends, the main menu scene replaces the intro scene
3. The main menu scene allows to select level scenes. Instead of replacing, the selected level scene will be pushed on top of the main menu scene. This pauses the main menu. The selected level scene is no overlay, therefore the main menu scene won't be updated and won't be rendered
4. Now it is possible to add a level pause scene on top. This scene is an overlay, therefore the level scene will be rendered below the pause scene
5. Popping the level scene from the stack will resume the main menu scene. Therefore the level scene doesn't have to know which scene will follow

```javascript
export class GameState {
    // […]
    update(dt) {} // Indirectly called by Kontra.
    render() {}   // Indirectly called by Kontra.
    onEnter() {}  // State entered. Reset the environment. Set listeners, …
    onExit() {}   // State left. Clean up the environment. Remove listeners, …
    onPause() {}  // Another state was set on top. Remove listeners, …
    onResume() {} // The state on top got removed. This is now the primary state. Add listeners, …
}
```

Kontra is calling `update` and `render` of the state machine.

```javascript
export class GameStateMachine {
    constructor() {
        /**
         * @type {GameState[]}
         */
        this.stateStack = [];
    }

    // Called by Kontra.
    // Updates the top most game state.
    // Could also update visible states but that's currently not required by my game.
    update(dt) {}

    // Called by Kontra.
    // Renders all visible game states (at least the top most state).
    render() {}

    /**
     * Pauses the current state and pushes the next state on top of the stack.
     * Afterwards the new state will be entered by calling {GameState.onEnter}.
     *
     * @param {GameState} state
     */
    pushState(state) {}
    popState() {}
    replaceState(state) {}
}
```

The code will be shared later.

**Comment by sdfereday** [comment]
I like it a lot, the idea of stacking up states looks pretty powerful. Right now I'm just handling some real simple stuff which will no doubt lose flexibility fairly soon (already starting to see it). Thanks for sharing this though @Rayne , it's definitely given more to think about in a positive way.

**Comment by Rayne** [comment]
My solution is probably a little over-engineered. Here are the relevant bits and pieces:

- [`GameStateMachine`](https://github.com/Rayne/js13k-lost-robot/blob/894149a412d65aab0bfac7313ec383be16ea45e4/public/src/GameState/GameStateMachine.js) handles game states and transitions

- The state interface is defined in [`GameState`](https://github.com/Rayne/js13k-lost-robot/blob/894149a412d65aab0bfac7313ec383be16ea45e4/public/src/GameState/GameState.js)

- **Example:** Pushing a `PauseMenuGameState` object to the state stack will pause the states below. `PauseMenuGameState` is an overlay that is popping itself from the state stack.

  https://github.com/Rayne/js13k-lost-robot/blob/894149a412d65aab0bfac7313ec383be16ea45e4/public/src/GameState/PauseMenuGameState.js#L10

  When the user is selecting the quit action, then a custom kontra event is triggered. That event is received by the loaded map, which will pop itself from the stack. This will resume the level select menu.

  https://github.com/Rayne/js13k-lost-robot/blob/894149a412d65aab0bfac7313ec383be16ea45e4/public/src/GameState/PauseMenuGameState.js#L39-L40

  The code above this snippet is not relevant for game state (or scene) handling. That code defines the transitions of another finite state machine ([`MenuStateMachine`](https://github.com/Rayne/js13k-lost-robot/blob/894149a412d65aab0bfac7313ec383be16ea45e4/public/src/Menu/MenuStateMachine.js)), that is controlling menu states. This allows the player to navigate menus with keyboard, mouse and touch controls

**Comment by straker** [comment]
Closed via #169
